src/dotty/tools/dotc/ast/untpd.scala

```diff
@@ -94,17 +94,51 @@ object untpd extends Trees.Instance[Untyped] with UntypedTreeInfo {
   class InfixOpBlock(leftOperand: Tree, rightOp: Tree) extends Block(leftOperand :: Nil, rightOp)
 
   // ----- Modifiers -----------------------------------------------------
+  /** Mod is intended to record syntactic information about modifiers, it's
+    * NOT a replacement of flagsets.
+    *
+    * For any query about semantic information, check `flags` instead.
+    */
+  sealed trait Mod extends Positioned
 
- /** Modifiers and annotations for definitions
-   *  @param flags          The set flags
+  object Mod {
+    case class Private() extends Mod
```

Comment:
@odersky I just checked the parser code, whether a semantic flag gets added depends on some compatibility check (see `addFlag`). But the syntactic `Mod` should be always added.

It's also possible to refactor the flag compatibility check to `Modifiers`, but the error reporting depends on compiler state (`lastErrorOffset`), make the refactoring difficult.